pkg/drivers/docker_driver.go

```diff
@@ -0,0 +1,371 @@
+	}
+
+	_, err = d.cli.WaitContainer(container.ID)
+
+	image, err := d.cli.CommitContainer(docker.CommitContainerOptions{
+		Container: container.ID,
+	})
+
+	if err != nil {
+		logrus.Errorf("Error committing container: %s", err.Error())
+	}
+
+	if !d.save {
+		if err = d.cli.RemoveContainer(docker.RemoveContainerOptions{
+			ID: container.ID,
+		}); err != nil {
+			logrus.Infof("Error when removing container %s: %s", container.ID, err.Error())
+		}
+	}
+
+	d.currentImage = image.ID
+	return image.ID
+}
+
+func (d *DockerDriver) exec(env []string, command []string) (string, string, int) {
+	// first, start container from the current image
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Env:          env,
+			Cmd:          command,
+			Entrypoint:   []string{},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+	})
+	if err != nil {
+		logrus.Errorf("Error creating container: %s", err.Error())
+		return "", "", -1
+	}
+	defer d.removeContainer(container.ID)
+
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+
+	if err = d.cli.StartContainer(container.ID, nil); err != nil {
+		logrus.Errorf("Error creating container: %s", err.Error())
+	}
+
+	//TODO(nkubala): look into adding timeout
+	exitCode, err := d.cli.WaitContainer(container.ID)
+	if err != nil {
+		logrus.Errorf("Error when waiting for container: %s", err.Error())
+	}
+
+	if err = d.cli.Logs(docker.LogsOptions{
+		Container:    container.ID,
+		OutputStream: stdout,
+		ErrorStream:  stderr,
+		Stdout:       true,
+		Stderr:       true,
+	}); err != nil {
+		logrus.Errorf("Error retrieving container logs: %s", err.Error())
+	}
+
+	return stdout.String(), stderr.String(), exitCode
+}
+
+func (d *DockerDriver) GetConfig() (unversioned.Config, error) {
+	img, err := d.cli.InspectImage(d.currentImage)
+	if err != nil {
+		logrus.Errorf("Error when inspecting image: %s", err.Error())
+		return unversioned.Config{}, err
+	}
+
+	// docker provides these as maps (since they can be mapped in docker run commands)
+	// since this will never be the case when built through a dockerfile, we convert to list of strings
+	volumes := []string{}
+	for v := range img.Config.Volumes {
+		volumes = append(volumes, v)
+	}
+
+	ports := []string{}
+	for p := range img.Config.ExposedPorts {
+		ports = append(ports, p.Port())
+	}
+
+	return unversioned.Config{
+		Env:          convertEnvToMap(img.Config.Env),
+		Entrypoint:   img.Config.Entrypoint,
+		Cmd:          img.Config.Cmd,
+		Volumes:      volumes,
+		Workdir:      img.Config.WorkingDir,
+		ExposedPorts: ports,
+	}, nil
+}
+
+func (d *DockerDriver) removeContainer(containerID string) {
+	if d.save {
+		return
+	}
+	if err := d.cli.RemoveContainer(docker.RemoveContainerOptions{
+		ID: containerID,
+	}); err != nil {
+		logrus.Infof("Error when removing container %s: %s", containerID, err.Error())
```

Comment:
warn

Reply:
I think a better pattern here is to have the functions return errors and their callers decide if they want to warn or error when checking the error. Otherwise it gets pretty confusing when and where errors are suppressed.

Reply:
yep, did this